src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -412,102 +534,81 @@ protected virtual void ValidateAtHash(OpenIdConnectProtocolValidationContext val
         /// <exception cref="OpenIdConnectProtocolInvalidNonceException">if the 'nonce' found in the 'id_token' does not match <see cref="OpenIdConnectProtocolValidationContext.Nonce"/>.</exception>
         /// <exception cref="OpenIdConnectProtocolInvalidNonceException">if <see cref="RequireTimeStampInNonce"/> is true and a timestamp is not: found, well formed, negatire or expired.</exception>
         /// <remarks>The timestamp is only validated if <see cref="RequireTimeStampInNonce"/> is true.
         /// <para>If <see cref="OpenIdConnectProtocolValidationContext.Nonce"/> is not-null, then a matching 'nonce' must exist in the 'id_token'.</para></remarks>
-        protected virtual void ValidateNonce(OpenIdConnectProtocolValidationContext validationContext)
+        protected virtual void ValidateNonce(JwtSecurityToken idToken, OpenIdConnectProtocolValidationContext validationContext)
         {
-            IdentityModelEventSource.Logger.WriteVerbose(LogMessages.IDX10342);
+            IdentityModelEventSource.Logger.WriteVerbose(LogMessages.IDX10319);
 
-            if (validationContext == null)
+            if (idToken == null)
             {
-                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10000, GetType() + ": validationContext"), typeof(ArgumentNullException), EventLevel.Verbose);
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10000, GetType() + ": idToken"), typeof(ArgumentNullException), EventLevel.Verbose);
             }
 
-            if (validationContext.IdToken == null)
+            if (validationContext == null)
             {
-                IdentityModelEventSource.Logger.WriteInformation(LogMessages.IDX10343);
-                return;
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10000, GetType() + ": validationContext"), typeof(ArgumentNullException), EventLevel.Verbose);
             }
 
-            string nonceFoundInJwt = validationContext.IdToken.Payload.Nonce;
-            if (RequireNonce)
-            {
-                if (validationContext.Nonce == null)
-                {
-                    LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10311), typeof(OpenIdConnectProtocolInvalidNonceException), EventLevel.Error);
-                }
-
-                if (nonceFoundInJwt == null)
-                {
-                    LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10322, validationContext.IdToken), typeof(OpenIdConnectProtocolInvalidNonceException), EventLevel.Error);
-                }
-            }
-            else if ((validationContext.Nonce != null) && (nonceFoundInJwt == null))
+            string nonceFoundInJwt = idToken.Payload.Nonce;
+            if (RequireNonce && validationContext.Nonce == null)
```

Comment:
There is a slight change in the logic that we should think about. If either validationContext.Nonce OR id_token.Nonce is non-null, then we need to check. This new logic seems to saw, if (RequireNonce == false then validationContext.Nonce == null && id_token.Nonce != null is OK. I don't think it is OK.